File: nuxt_side/components/Gallery/models/recipe_info.js
import MemoryRecord from 'js-memory-record'

export class RecipeInfo extends MemoryRecord {
  static get field_label() {
    return "出力フォーマット"
  }

  static get field_message() {
    return ""
  }

  static get define() {
    return [
      // ../../../app/models/recipe_info.rb
      { key: "is_recipe_mp4",  name: "MP4",  real_ext: "mp4",  loop_key_enable: false,  file_type: "video", environment: ["development", "staging", "production"], title: "一般的な動画形式",                  message: "◎Twitter ◎画質 ◎シーク ◎BGM", media_p: true,  thumbnail_p: true,  },
      { key: "is_recipe_gif",  name: "GIF",  real_ext: "gif",  loop_key_enable: true,   file_type: "image", environment: ["development", "staging", "production"], title: "一般的な画像切替形式",              message: "◎Twitter ◎画質 ×シーク",       media_p: true,  thumbnail_p: false, },
      { key: "is_recipe_apng", name: "APNG", real_ext: "apng", loop_key_enable: true,   file_type: "image", environment: ["development", "staging", "production"], title: "GIFの代替として注目されていた",     message: "×Twitter ◎画質 ×シーク",       media_p: true,  thumbnail_p: false, },
      { key: "is_recipe_webp", name: "WebP", real_ext: "webp", loop_key_enable: true,   file_type: "image", environment: ["development", "staging", "production"], title: "Googleだけが推している",            message: "×Twitter ◎画質 ×シーク",       media_p: true,  thumbnail_p: false, },
      { key: "is_recipe_png",  name: "PNG",  real_ext: "png",  loop_key_enable: false,  file_type: "image", environment: ["development", "staging", "production"], title: "最後の局面の画像",                  message: "1枚だけなので出来上がるのが速い", media_p: true,  thumbnail_p: false, },
      { key: "is_recipe_zip",  name: "ZIP",  real_ext: "zip",  loop_key_enable: false,  file_type: "zip",   environment: ["development", "staging", "production"], title: "PNG画像の詰め合わせ",               message: "自分で何かしたい人向け",          media_p: false, thumbnail_p: false, },
      // { key: "is_recipe_mov",             name: "MOV",               real_ext: "mov",  loop_key_enable: false, file_type: "video", environment: ["development"],                          message: "PC(WEB)から投稿できない。モバイルアプリから投稿できる",   },
      // { key: "is_recipe_jpg",             name: "JPG",               real_ext: "jpg",  loop_key_enable: false, file_type: "image", environment: ["development"],                          message: "最後の局面の画像",                                        },
      // { key: "is_recipe_bmp",             name: "BMP",               real_ext: "bmp",  loop_key_enable: false, file_type: "image", environment: ["development"],                          message: "最後の局面の画像",                                        },
      // { key: "is_recipe_mp4_yuv444p10le", name: "MP4 YUV444",        real_ext: "mp4",  loop_key_enable: false, file_type: "video", environment: ["development"],                          message: "PC(WEB)から投稿できない。モバイルアプリからは投稿できる", },
      // { key: "is_recipe_mp4_noopt",       name: "MP4 (noopt)",       real_ext: "mp4",  loop_key_enable: false, file_type: "video", environment: ["development"],                          message: "PC(WEB)から投稿できる。モバイルアプリからも投稿できる",   },
    ]
  }
}

File: app/models/recipe_info.rb
# TWITTERに投稿できる動画と画像の仕様について ← めちゃくちゃ詳しい
# https://nico-lab.net/twitter_upload_format_spec/
#
# 色の情報の扱いについて
# https://nico-lab.net/setting_in_out_color_with_ffmpeg/
#
# TwitterにMP4動画をアップロードするにはyuv420pを使う必要がある（らしい）
# https://kivantium.hateblo.jp/entry/2017/07/16/160859
#
# ・yuv420p にすることで QuickTime で再生できる
# ・yuv420p にすることで Twitter に投稿できる
#

class RecipeInfo
  include ApplicationMemoryRecord
  memory_record [
    # ../../nuxt_side/components/kiwi/models/recipe_info.js
    { key: "is_recipe_mp4",               name: "MP4",                  real_ext: "mp4",  to_method: "to_animation_mp4",  media_p: true,  thumbnail_p: true, og_video_p: true, override_options: {},                        },
    { key: "is_recipe_gif",               name: "GIF",                  real_ext: "gif",  to_method: "to_animation_gif",  media_p: true,  thumbnail_p: false, og_video_p: false, override_options: {},                   },
    { key: "is_recipe_apng",              name: "APNG",                 real_ext: "apng", to_method: "to_animation_apng", media_p: true,  thumbnail_p: false, og_video_p: false, override_options: {},                      },
    { key: "is_recipe_webp",              name: "WebP",                 real_ext: "webp", to_method: "to_animation_webp", media_p: true,  thumbnail_p: false, og_video_p: false, override_options: {},                      },
    { key: "is_recipe_png",               name: "PNG",                  real_ext: "png",  to_method: "to_png",            media_p: true,  thumbnail_p: false, og_video_p: false, override_options: {},                      },
    { key: "is_recipe_zip",               name: "ZIP",                  real_ext: "zip",  to_method: "to_animation_zip",  media_p: false,  thumbnail_p: false, og_video_p: false, override_options: {},                      },
    # { key: "is_recipe_mov",             name: "MOV",                  real_ext: "mov",  to_method: "animation",         media_p: true,  thumbnail_p: false, og_video_p: false, override_options: {},                        },
    # { key: "is_recipe_jpg",             name: "JPG",                  real_ext: "jpg",  to_method: "image",             media_p: true,  thumbnail_p: false, og_video_p: false, override_options: {},                      },
    # { key: "is_recipe_bmp",             name: "BMP",                  real_ext: "bmp",  to_method: "image",             media_p: true,  thumbnail_p: false, og_video_p: false, override_options: {},                      },
    # { key: "is_recipe_mp4_yuv444p10le", name: "MP4 (yuv444p10le)",    real_ext: "mp4",  to_method: "animation",         media_p: true,  thumbnail_p: false, og_video_p: false, override_options: {},                      },
    # { key: "is_recipe_mp4_noopt",       name: "MP4 (noopt)",          real_ext: "mp4",  to_method: "animation",         media_p: true,  thumbnail_p: false, og_video_p: false, override_options: {optimize_layer: false},   },
  ]
end
